common/src/stack/ansible/plugins/inventory/stacki.py

```diff
@@ -0,0 +1,130 @@
+# @copyright@
+# Copyright (c) 2006 - 2020 Teradata
+# All rights reserved. Stacki(r) v5.x stacki.com
+# https://github.com/Teradata/stacki/blob/master/LICENSE.txt
+# @copyright@
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+    inventory: stacki
+    short_description: Returns hosts, groups, and vars from Stacki
+    description:
+        - Reads hosts info and attributes from Stacki.
+        - Returns host groups based on standard Stacki host selectors.
+        - Hosts have attributes attached as vars.
+"""
+
+EXAMPLES = """
+# Directly target hosts
+ansible backend-0-0 -m ping
+
+# Using standard host selectors
+ansible a_backend -m ping
+
+# In Ansible playbooks, you can use the selectors in the `hosts` attribute:
+- name: Ping my backends
+  hosts: a_backend
+  tasks:
+    - ping:
+
+# You can also use the --limit flag to `ansible-playbook`
+ansible-playbook --limit b_default playbook.yml
+
+# The following types of groups are created, with the UPPERCASE replaced
+# with the Stacki primitive:
+- a_APPLIANCE
+- e_ENVIRONMENT
+- o_OS
+- b_BOX
+- g_GROUP
+- r_RACK
+
+# All the attributes for a host are available as Ansible host vars:
+- name: Ping managed hosts
+  hosts: all
+  tasks:
+    - ping:
+      when: managed
+
+# To list all the inventory hosts, groups, and vars run:
+ansible-inventory --list
+"""
+
+from collections import defaultdict
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.plugins.inventory import BaseInventoryPlugin
+import stack.api
+
+
+class InventoryModule(BaseInventoryPlugin):
+	NAME = "stacki"
+
+	def verify_file(self, path):
+		return path.endswith("@stacki")
+
+	def parse(self, inventory, loader, path, cache=True):
```

Comment:
This function is fairly large and might benefit from being broken into smaller pieces. There appears to be two or three blocks that could easily be encapsulated in other functions.

This subdivision could aid in potential unit testing to increase code coverage/hit otherwise hard to test error cases.

Reply:
It's 40 lines of code. Wear your big boy pants and don't complicate things.